File: app/src/main/java/com/pietrantuono/podcasts/addpodcast/model/AddPodcastsModelImpl.java
package com.pietrantuono.podcasts.addpodcast.model;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.observers.DisposableObserver;

public class AddPodcastsModelImpl implements AddPodcastsModel {
    private final SearchApi searchApi;
    Disposable subscription;
    private DisposableObserver<SearchResult> observer;
    Observable<SearchResult> cachedRequest;

    public AddPodcastsModelImpl(SearchApi searchApi) {
        this.searchApi = searchApi;
    }

    @Override
    public void subscribeToSearch(DisposableObserver<SearchResult> observer) {
        this.observer = observer;
        if (cachedRequest != null) {
            subscription = cachedRequest.subscribeWith(observer);
        }
    }

    @Override
    public void unsubscribeFromSearch() {
        if (subscription != null && !subscription.isDisposed()) {
            subscription.dispose();
        }
    }

    @Override
    public void searchPodcasts(String query) {
        cachedRequest = searchApi.search(query);
        subscription = cachedRequest.subscribeWith(observer);
    }

}

File: app/src/main/java/com/pietrantuono/podcasts/addpodcast/model/AddPodcastsModelImpl.kt
package com.pietrantuono.podcasts.addpodcast.model

import com.pietrantuono.podcasts.addpodcast.singlepodcast.view.State
import rx.Observable
import rx.Observer
import rx.Subscription

class AddPodcastsModelImpl(private val searchApi: SearchApi) : AddPodcastsModel {
    internal var subscription: Subscription? = null
    private var observer: Observer<SearchResult>? = null
    internal var cachedRequest: Observable<SearchResult>? = null
    override var viewState: State? = null
    override var query: String? = null

    override fun subscribeToSearch(observer: Observer<SearchResult>) {
        this.observer = observer
        subscription = cachedRequest?.subscribe(observer)
    }

    override fun unsubscribeFromSearch() {
        if (subscription?.isUnsubscribed == false) {
            subscription?.unsubscribe()
        }
    }

    override fun searchPodcasts(query: String?) {
        this.query = query
        cachedRequest = searchApi.search(query)
        subscription = cachedRequest?.subscribe(observer)
    }

    override fun refresh() {
        query?.let { searchPodcasts(it) }
    }
}
